### PR TITLE
Refactor checkGlobals() error message creation

### PR DESCRIPTION
**Pull request by plroebuck**
### Description of the Change
Modified to better deal with diagnostic messages dealing with plurality, and markup of output.

### Alternate Designs
Routines are simplified versions inspired by their [R counterparts](https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#Diagnostic-messages).

### Why should this be in core?
NA

### Benefits
Codebase better situated if i18n ever considered.
Routines now available for use elsewhere.

### Possible Drawbacks
None to knowledge.

### Applicable issues
semver-patch

### CODE REVIEW

**Comment by craigtaub** [comment]
Looks good. Can you add tests for them?

File: test/unit/runner.spec.js [hunk 1]
```diff
@@ -103,9 +103,9 @@ describe('Runner', function() {
       var test = new Test('im a test', noop);
       runner.checkGlobals();
       global.foo = 'bar';
-      runner.on('fail', function(_test, err) {
```
[the remark this answers is not in the document]
Comment (plroebuck) [reply test/unit/runner.spec.js]: No. This PR came first. While working with this one, I noticed the shadowing and fixed related area. When I looked harder, I saw the problem was more widespread and created the shadowing fix PR.

File: lib/utils.js [hunk 2]
```diff
@@ -736,6 +736,76 @@ exports.clamp = function clamp(value, range) {
...
+ *   ngettext(
+ *     pkgs.length,
+ *     'cannot load package: %s',
+ *     'cannot load packages: %s'
+ *   ),
+ *   pkgs.map(sQuote).join(', ')
+ * );
+ * console.log(msg); // => cannot load packages: 'one', 'two'
+ */
+exports.ngettext = function(n, msg1, msg2) {
```
[the remark this answers is not in the document]
Comment (plroebuck) [reply lib/utils.js]: IMO, `plural` takes us in the wrong direction. Similar conceptionally, except it evaluates dozens of rules to determine what the plural spelling _would_ be in English -- something I already know.

File: lib/utils.js [hunk 3]
```diff
@@ -565,7 +566,11 @@ exports.lookupFiles = function lookupFiles(filepath, extensions, recursive) {
     }
     if (!extensions) {
       throw createMissingArgumentError(
-        'Argument "extensions" required when argument "filepath" is a directory',
+        util.format(
+          'Argument %s required when argument %s is a directory',
+          exports.sQuote('extensions'),
+          exports.sQuote('filepath')
+        ),
```
Comment (juergba) [inline lib/utils.js]: Does this work, single quotes within single quotes?

Comment (plroebuck) [reply lib/utils.js]: Mais bien sûr!

**Comment by plroebuck** [comment]
> I'm worried a little bit, we have ESLint, prettier/prettier, possibly even more Lint (lexical order) ...

Not following this comment. Nothing from this PR changes those tools, or is any more or less impacted by them...

> In some of your changes here, I personally would just set those quotes manually.

Fair. I find it easier to read not having the noise of a manually formatted one.

If string has multiple cases, I find `util.format` (or `sprintf`) far easier to read and maintain.
But even for a single case, the manually quoted version requires three fragments; this style eliminates one of those fragments.

**Comment by plroebuck** [comment]
anyone else? otherwise, will commit tomorrow.

File: lib/runner.js [hunk 4]
```diff
@@ -204,13 +208,14 @@ Runner.prototype.checkGlobals = function(test) {
...
-      new Error('global leaks detected: ' + leaks.join(', ') + '')
+  if (leaks.length) {
+    var format = ngettext(
+      leaks.length,
+      'global leak detected: %s',
+      'global leaks detected: %s'
     );
-  } else if (leaks.length) {
-    this.fail(test, new Error('global leak detected: ' + leaks[0]));
+    var error = new Error(util.format(format, leaks.map(sQuote).join(', ')));
```
[the remark this answers is not in the document]
Comment (cyjake) [reply lib/runner.js]: [template literal](https://node.green/#ES2015-syntax-template-literals) is available in `node >= 6`, which matches the version range mocha now supports.

[the remark this answers is not in the document]
Comment (boneskull) [reply lib/runner.js]: @cyjake much of the codebase can't use ES6 features because it runs in IE11.

[the remark this answers is not in the document]
Comment (cyjake) [reply lib/runner.js]: ah, i see. What about adding babel to browserify process? Anyway, I shouldn't propose these kinds of change here. It's good to move forward though.

**Pull request merged**